server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MDNSendMethod.scala

```diff
@@ -0,0 +1,316 @@
+    }
+
+  private def buildMailAndResponse(identity: Identity, sender: String, requestEntry: MDNSendCreateRequest, originalMessage: Message): Either[Exception, (MailImpl, MDNSendCreateResponse)] =
+    for {
+      mailRecipient <- getMailRecipient(originalMessage)
+      mdnFinalRecipient <- getMDNFinalRecipient(requestEntry, identity)
+      mdn = buildMDN(requestEntry, originalMessage, mdnFinalRecipient)
+      subject = buildMessageSubject(requestEntry, originalMessage)
+      (mailImpl, mimeMessage) = buildMailAndMimeMessage(sender, mailRecipient, subject, mdn)
+    } yield {
+      (mailImpl, buildMDNSendCreateResponse(requestEntry, mdn, mimeMessage))
+    }
+
+  private def buildMailAndMimeMessage(sender: String, recipient: String, subject: String, mdn: MDN): (MailImpl, MimeMessage) = {
+    val mimeMessage: MimeMessage = mdn.asMimeMessage()
+    mimeMessage.setFrom(sender)
+    mimeMessage.setRecipients(javax.mail.Message.RecipientType.TO, recipient)
+    mimeMessage.setSubject(subject)
+    mimeMessage.saveChanges()
+
+    val mailImpl: MailImpl = MailImpl.builder()
+      .name(MDNId.generate.value)
+      .sender(sender)
+      .addRecipient(recipient)
+      .mimeMessage(mimeMessage)
+      .build()
+    mailImpl -> mimeMessage
+  }
+
+  private def getMailRecipient(originalMessage: Message): Either[MDNSendNotFoundException, String] =
+    originalMessage.getHeader.getFields(DISPOSITION_NOTIFICATION_TO)
+      .asScala
+      .headOption
+      .map(field => AddressListFieldLenientImpl.PARSER.parse(field, new DecodeMonitor))
+      .map(addressListField => addressListField.getAddressList)
+      .map(addressList => addressList.flatten())
+      .flatMap(mailboxList => mailboxList.stream().findAny().toScala)
+      .map(mailbox => mailbox.getAddress)
+      .toRight(MDNSendNotFoundException("Invalid \"Disposition-Notification-To\" header field."))
+
+  private def getMDNFinalRecipient(requestEntry: MDNSendCreateRequest, identity: Identity): Either[MDNSendForbiddenFromException, FinalRecipient] =
+    requestEntry.finalRecipient
+      .map(finalRecipient => finalRecipient.getMailAddress)
+      .map(mayBeMailAddress => (mayBeMailAddress.isSuccess && mayBeMailAddress.get.equals(identity.email)))
+      .map {
+        case true => scala.Right(requestEntry.finalRecipient.get.asJava.get)
+        case false => Left(MDNSendForbiddenFromException("The user is not allowed to use the given \"finalRecipient\" property"))
+      }
+      .getOrElse(scala.Right(FinalRecipient.builder()
+        .finalRecipient(Text.fromRawText(identity.email.asString()))
+        .build()))
+
+  private def buildMDN(requestEntry: MDNSendCreateRequest, originalMessage: Message, finalRecipient: FinalRecipient): MDN = {
+    val reportBuilder: MDNReport.Builder = MDNReport.builder()
+      .dispositionField(requestEntry.disposition.asJava.get)
+      .finalRecipientField(finalRecipient)
+      .originalRecipientField(originalMessage.getTo.asScala.head.toString)
```

Comment:
I don't feel sure about this. 
What happened if `originalMessage.getTo` has multi-element?

Reply:
`.originalRecipientField(mailboxSession.getUser().asString())`

From the draft we use the current user. We should be doing the same here IMO.